contrib/mesos/pkg/minion/task.go

```diff
@@ -0,0 +1,519 @@
+			if exitError, ok := err.(exitError); ok {
+				if waitStatus, ok := exitError.Sys().(syscall.WaitStatus); ok {
+					wr.completed = &taskCompletion{
+						name: t.name,
+						code: waitStatus.ExitStatus(),
+					}
+					return
+				}
+			}
+			wr.err = fmt.Errorf("task wait ended strangely for %q: %v", t.bin, err)
+		} else {
+			wr.completed = &taskCompletion{name: t.name}
+		}
+	}()
+
+	// wait for the process to complete, or else for a "quit" signal on the task (at which point we'll attempt to kill manually)
+	select {
+	case <-t.shouldQuit:
+		// check for tie
+		select {
+		case wr := <-waitCh:
+			// we got a signal to quit, but we're already finished; attempt best effort delvery
+			trySend(wr)
+		default:
+			// Wait() has not exited yet, kill the process
+			log.Infof("killing %s : %s", t.name, t.bin)
+			pid, err := t.cmd.Kill()
+			if err != nil {
+				t.tryError(fmt.Errorf("failed to kill process: %q pid %d: %v", t.bin, pid, err))
+			}
+		}
+	case wr := <-waitCh:
+		// task has completed before we were told to quit, pass along completion and error information
+		trySend(wr)
+	}
+	return taskShouldRestart
+}
+
+func forwardCompletionUntil(tin <-chan *taskCompletion, tch chan<- *taskCompletion, discard <-chan struct{}, abort <-chan struct{}, tap func(*taskCompletion, bool)) {
+	var tc *taskCompletion
+	var ok bool
+forwardLoop:
+	for {
+		select {
+		case tc, ok = <-tin:
+			if !ok {
+				return
+			}
+			if tap != nil {
+				tap(tc, false)
+			}
+			select {
+			case <-abort:
+				break forwardLoop
+			case <-discard:
+			case tch <- tc:
+			}
+		case <-abort:
+			// best effort
+			select {
+			case tc, ok = <-tin:
+				if ok {
+					if tap != nil {
+						tap(tc, true)
+					}
+					break forwardLoop
+				}
+			default:
+			}
+			return
+		}
+	}
+	// best effort
+	select {
+	case tch <- tc:
+	case <-discard:
+	default:
+	}
+}
+
+func forwardErrorsUntil(ein <-chan error, ech chan<- error, discard <-chan struct{}, abort <-chan struct{}, tap func(error, bool)) {
+	var err error
+	var ok bool
+forwardLoop:
+	for {
+		select {
+		case err, ok = <-ein:
+			if !ok {
+				return
+			}
+			if tap != nil {
+				tap(err, false)
+			}
+			select {
+			case <-abort:
+				break forwardLoop
+			case <-discard:
+			case ech <- err:
+			}
+		case <-abort:
+			// best effort
+			select {
+			case err, ok = <-ein:
+				if ok {
+					if tap != nil {
+						tap(err, true)
+					}
+					break forwardLoop
+				}
+			default:
+			}
+			return
+		}
+	}
+	// best effort
+	select {
+	case ech <- err:
+	case <-discard:
+	default:
+	}
+}
+
+// forwardUntil forwards task process completion status and errors to the given output
+// chans until either the task terminates or abort is closed.
+func (t *task) forwardUntil(tch chan<- *taskCompletion, ech chan<- error, abort <-chan struct{}) {
+	// merge task completion and error until we're told to die, then
+	// tell the task to stop
+	defer close(t.shouldQuit)
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
```

Comment:
with taskCompletion and error merged into one type or interface, this function would be a two-liner.